tools/codegen/gen_functionalization_type.py

```diff
@@ -444,41 +478,78 @@ def emit_registration_helper(f: NativeFunction, *, is_view: bool) -> str:
-        view_str = [emit_registration_helper(g.view, is_view=True)]
-        if g.view_inplace is not None:
-            view_str.append(emit_registration_helper(g.view_inplace, is_view=True))
-        return view_str
+        # we always want to register composite kernels
+        if g.view.has_composite_implicit_autograd_kernel:
+            view_str = [emit_registration_helper(g.view)]
+            if g.view_inplace is not None:
+                assert g.view_inplace.is_view_op
+                # We don't currently have any composite view ops that have a non-composite
+                # "inplace_view" variant, and implementing the logic it is a little annoying.
+                assert g.view_inplace.has_composite_implicit_autograd_kernel
+                view_str.append(emit_registration_helper(g.view_inplace))
+            return view_str
+
+        # functionalization needs to register kernels for view + view_inplace ops
+        if funcTarget == FunctionalizationTarget.FUNCTIONALIZATION:
+            view_str = [emit_registration_helper(g.view)]
+            if g.view_inplace is not None:
+                assert g.view_inplace.is_view_op
+                view_str.append(emit_registration_helper(g.view_inplace))
+            return view_str
+
+        # while "add back view" kernels need to register kernels for view_copy ops
+        assert funcTarget == FunctionalizationTarget.ADD_BACK_VIEWS
+        assert g.view_copy is not None
+        return [emit_registration_helper(g.view_copy)]
     else:
         f = g
-        return [emit_registration_helper(f, is_view=False)]
+        assert not f.is_view_op
+        return [emit_registration_helper(f)]
 
 def gen_functionalization_definition(
     selector: SelectiveBuilder,
     g: Union[NativeFunction, NativeFunctionsViewGroup],
-    functional_op: Optional[NativeFunction]
+    functional_op: Optional[NativeFunction],
+    funcTarget: FunctionalizationTarget,
 ) -> List[str]:
-    if not needs_functionalization(selector, g):
+    if not needs_functionalization(selector, g, Target.ANONYMOUS_DEFINITION, funcTarget):
         return []
-    if isinstance(g, NativeFunctionsViewGroup):
-        view_defs = []
-        if not g.view.has_composite_implicit_autograd_kernel:
-            # For now I'm just asserting this - later, the view functionalization kernels
-            # should be updated to redispatch to view_copy().
-            assert g.view_copy is not None
-            view_defs.append(emit_view_functionalization_body(g.view, g.view))
-        if g.view_inplace is not None and not g.view_inplace.has_composite_implicit_autograd_kernel:
-            view_defs.append(emit_view_functionalization_body(g.view_inplace, g.view))
-        return view_defs
-    else:
-        f = g
-        assert modifies_arguments(f)
-        return [emit_inplace_functionalization_body(f, functional_op)]
+    if funcTarget == FunctionalizationTarget.FUNCTIONALIZATION:
```

Comment:
it's more idiomatic to do these conditions as

```
if funcTarget is FunctionalizationTarget.FUNCTIONALIZATION:
  ...
elif funcTarget is FunctionalizationTarget.ADD_BACK_VIEWS:
  ...
else:
  assert_never(funcTarget)
```

you then get compile time exhaustiveness checking

Reply:
ah yep - this will go away though if I'm using TLS to keep the views around instead of an extra dispatch key / set of kernels to generate.